File: javascript/grep/grep.js
#!/usr/bin/env node

const fs = require('fs');

const opts = {
  lineNumbers:     false, // -n
  caseInsensitive: false, // -i
  fileNameOnly:    false, // -l
  wholeLine:       false, // -x
  inverted:        false, // -v
};

function main() {
  let args = parseOptions(process.argv.slice(2));
  if (args.length < 2) {
    console.error('usage: grep.js [options] pattern file [...]');
    process.exit(1);
  }
  let pattern = getPattern(args.shift());
  args.forEach(filename => {
    processFile(filename, pattern, args.length);
  });
}

function parseOptions(args) {
  while (args[0].startsWith('-')) {
    let option = args.shift();
    switch (option) {
      case '-n':
        opts.lineNumbers = true;
        break;
      case '-i':
        opts.caseInsensitive = true;
        break;
      case '-l':
        opts.fileNameOnly = true;
        break;
      case '-x':
        opts.wholeLine = true;
        break;
      case '-v':
        opts.inverted = true;
        break;
      default:
        console.error('Error: unknown option ' + option);
        process.exit(1);
    }
  }
  return args;
}

function getPattern(patternString) {
  if (opts.wholeLine) {
    patternString = '^' + patternString + '$';
  }
  return new RegExp(patternString, opts.caseInsensitive ? 'i' : '');
}

function processFile(filename, pattern, numFiles) {
  // TODO: validate file exists and is readable
  const lines = fs.readFileSync(filename, 'utf8').split('\n');

  // XOR is a bitwise operator that requires ints, not booleans.
  const inv = opts.inverted ? 1 : 0;

  let lineNum = 1;
  for (const line of lines) {
    const match = line.match(pattern) ? 1 : 0;
    if (inv ^ match) {
      emit(filename, line, lineNum, numFiles);
      if (opts.fileNameOnly) {
        break;
      }
    }
    lineNum += 1;
  }
}

function emit(filename, line, lineNum, numFiles) {
  const output = [];
  if (opts.fileNameOnly) {
    output.push(filename);
  }
  else {
    if (numFiles > 1) {
      output.push(filename);
    }
    if (opts.lineNumbers) {
      output.push(lineNum);
    }
    output.push(line);
  }
  console.log(output.join(':'));
}

main();

File: python/grep/grep.py
import re


def grep(pattern, flags, files):
    match, output_fmt, opt_l = process_args(pattern, files, flags)
    results = []
    for file in files:
        nr = 0
        with open(file) as f:
            for line in f.readlines():
                nr += 1
                line = line.rstrip("\n")
                if match(line):
                    results.append(
                        output_fmt.format(F=file, N=nr, L=line)
                    )
                    if opt_l:
                        break

    return "".join(results)


def process_args(pattern, files, flags):
    opts = flags.split()

    # how to match the pattern against the line
    if '-x' in opts:
        pattern = f'^{pattern}$'
    
    re_opt = re.I if '-i' in opts else 0
    patt = re.compile(pattern, re_opt)
    match = lambda line: re.search(patt, line)

    if '-v' in opts:
        pos_match = match
        match = lambda line: not pos_match(line)

    # control the output:
    # F = filename, N = line number, L = line contents
    if '-l' in opts:
        fmt = "{F}"
    elif len(files) > 1:
        fmt = "{F}:{N}:{L}" if '-n' in opts else "{F}:{L}"
    else:
        fmt = "{N}:{L}" if '-n' in opts else "{L}"
    fmt += "\n"

    return match, fmt, '-l' in opts
